internal/pubsub/awspubsub/awspubsub.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/google/go-cloud/internal/pubsub"
+	"github.com/google/go-cloud/internal/pubsub/driver"
+)
+
+type topic struct {
+	client *sns.SNS
+	arn    string
+}
+
+// TopicOptions will contain configuration for topics.
+type TopicOptions struct{}
+
+// OpenTopic opens the topic on AWS SNS for the given projectID and
+// topicName.
+func OpenTopic(ctx context.Context, client *sns.SNS, topicARN string, opts *TopicOptions) *pubsub.Topic {
+	dt := openTopic(ctx, client, topicARN)
+	return pubsub.NewTopic(dt)
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(ctx context.Context, client *sns.SNS, topicARN string) driver.Topic {
+	return &topic{client, topicARN}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	for _, dm := range dms {
+		b := string(dm.Body)
+		attrs := map[string]*sns.MessageAttributeValue{}
+		for k, v := range dm.Metadata {
+			attrs[k] = &sns.MessageAttributeValue{
+				DataType:    aws.String("String"),
+				StringValue: aws.String(v),
+			}
+		}
+		params := sns.PublishInput{
+			Message:           &b,
+			MessageAttributes: attrs,
+			TopicArn:          &t.arn,
+		}
+		req, _ := t.client.PublishRequest(&params)
+		if err := req.Send(); err != nil {
+			return fmt.Errorf("awspubsub: failed to send message: %v", err)
+		}
+	}
+	return nil
+}
+
+// IsRetryable implements driver.Topic.IsRetryable.
+func (t *topic) IsRetryable(error) bool {
+	// The client handles retries.
+	return false
+}
+
+// As implements driver.Topic.As.
+func (t *topic) As(i interface{}) bool {
+	c, ok := i.(**sns.SNS)
+	if !ok {
+		return false
+	}
+	*c = t.client
+	return true
+}
+
+type subscription struct {
+	client *sqs.SQS
+	qURL   string
+}
+
+// SubscriptionOptions will contain configuration for subscriptions.
+type SubscriptionOptions struct{}
+
+// OpenSubscription opens the queue on AWS SQS for the given
+// projectID and subscriptionName.
+func OpenSubscription(ctx context.Context, client *sqs.SQS, qURL string, opts *SubscriptionOptions) *pubsub.Subscription {
+	ds := openSubscription(ctx, client, qURL)
+	return pubsub.NewSubscription(ds)
+}
+
+// openSubscription returns a driver.Subscription.
+func openSubscription(ctx context.Context, client *sqs.SQS, qURL string) driver.Subscription {
+	return &subscription{client, qURL}
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	params := sqs.ReceiveMessageInput{QueueUrl: &s.qURL}
+	req, output := s.client.ReceiveMessageRequest(&params)
+	err := req.Send()
+	if err != nil {
+		return nil, fmt.Errorf("awspubsub: message receive failed: %v", err)
+	}
+	var ms []*driver.Message
+	for _, m := range output.Messages {
+		type Attribute struct {
+			Type  string
```

Comment:
Is this just unused? What are possible values (given that the `Value` is always of type `string`)?

Reply:
Oh, right, it's unused. Removed it.

Reply:
Possible values for Type are String, String.Array, Number, and Binary. awspubsub on Go Cloud only creates messages whose attribute values have type String for now but messages could be sent to the same topic not by Go Cloud.